Applications/Spire/Source/Ui/ColorSwatch.cpp

```diff
@@ -0,0 +1,87 @@
+#include "Spire/Ui/ColorSwatch.hpp"
+#include "Spire/Spire/Dimensions.hpp"
+#include "Spire/Spire/LocalValueModel.hpp"
+#include "Spire/Styles/ChainExpression.hpp"
+#include "Spire/Styles/LinearExpression.hpp"
+#include "Spire/Styles/RevertExpression.hpp"
+#include "Spire/Styles/TimeoutExpression.hpp"
+#include "Spire/Ui/Box.hpp"
+#include "Spire/Ui/Layouts.hpp"
+
+using namespace boost::posix_time;
+using namespace Spire;
+using namespace Spire::Styles;
+
+namespace {
+  auto DEFAULT_STYLE() {
+    auto style = StyleSheet();
+    style.get(Any()).
+      set(border(scale_width(1), QColor(Qt::transparent))).
+      set(horizontal_padding(scale_width(1))).
+      set(vertical_padding(scale_height(1)));
+    style.get(Hover() || Focus()).
+      set(border_color(QColor(0x4B23A0)));
+    style.get(Drag()).
+      set(BackgroundColor(QColor(255, 255, 255, 204))).
+      set(border_color(QColor(0x4B23A0)));
+    style.get(Any() >> Body()).
+      set(border(scale_width(1), QColor(0, 0, 0, 51)));
+    return style;
+  }
+}
+
+ColorSwatch::ColorSwatch(QWidget* parent)
+  : ColorSwatch(std::make_shared<LocalColorModel>(), parent) {}
+
+ColorSwatch::ColorSwatch(std::shared_ptr<ColorModel> current,
+    QWidget* parent)
+    : QWidget(parent),
+      m_current(std::move(current)),
+      m_is_highlighted(false) {
+  setFocusPolicy(Qt::StrongFocus);
+  auto swatch = new Box();
+  swatch->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  auto box = new Box(swatch);
+  enclose(*this, *box);
+  proxy_style(*this, *box);
+  set_style(*this, DEFAULT_STYLE());
+  on_current(m_current->get());
+  m_current_connection = m_current->connect_update_signal(
+    std::bind_front(&ColorSwatch::on_current, this));
+}
+
+const std::shared_ptr<ColorModel>& ColorSwatch::get_current() const {
+  return m_current;
+}
+
+bool ColorSwatch::is_highlighted() const {
+  return m_is_highlighted;
+}
+
+void ColorSwatch::set_highlighted(bool highlighted) {
+  if(highlighted == m_is_highlighted) {
+    return;
+  }
+  m_is_highlighted = highlighted;
+  if(m_is_highlighted) {
+    match(*this, Highlighted());
+  } else {
+    unmatch(*this, Highlighted());
+  }
+}
+
+void ColorSwatch::on_current(const QColor& color) {
+  update_style(*this, [&] (auto& style) {
+    style.get(Any() >> Body()).set(BackgroundColor(color));
+    style.get(Drag() >> Body()).
+      set(BackgroundColor(
+        QColor(color.red(), color.green(), color.blue(), 204)));
+    style.get(Highlighted() >> Body()).
```

Comment:
I don't see a use of ```color``` here. Could this not be moved to the default style?

Reply:
When both the Drag and Highlighted selectors are applied to Body(), the Highlighted style is obscured by the Drag style. From my experiment, it appears that the Highlighted style remains visible only when it is applied after the Drag style. Hence, it was placed here.

Reply:
Considering that highlighted style is rarely set while dragging, I don't think it would be an issue that the Highlighted style obscures the Drag style. I will update this code.